### PR TITLE
Clustering metrics and MLflow experiment tracking

### PR DESCRIPTION
**Pull request by hemajv**
1. Additional clustering metrics to evaluate and compare the clustering behavior of systems for two different days.   
- Modifications made to the codebase: https://github.com/ManageIQ/aiops-insights-clustering/pull/32/files#diff-d0301332bd6fef353ec35837646aa49e
- (Detailed notebook of the code can be found here: https://gitlab.cee.redhat.com/insights-aiops/modeling/blob/master/use_cases/clustering/metric_comparing.ipynb) 

2. Integrating the MLflow experiment tracking template for logging the clustering model metrics. (https://github.com/AICoE/experiment-tracking/tree/master/openshift). 
- Updated instructions in the README for running the clustering model as an mlflow job on OpenShift: https://github.com/ManageIQ/aiops-insights-clustering/pull/32/files#diff-04c6e90faac2675aa89e2176d2eec7d8
- Added a separate task in the Makefile for the mlflow experiment tracking:
https://github.com/ManageIQ/aiops-insights-clustering/pull/32/files#diff-b67911656ef5d18c4ae36cb6741b7965

(Reference: https://github.com/AICoE/experiment-tracking-template, a demo application which shows how this template is being used to run mlflow jobs on OpenShift: )

### CODE REVIEW

**Comment by durandom** [comment]
@hemajv please also add all modifications to the codebase and the Makefile to this PR.

**Comment by hemajv** [comment]
@durandom Added the changes to the codebase and Makefile.

File: metric_tracking/__init__.py [hunk 1]
```diff
@@ -15,12 +16,13 @@ def do_tracking():
         mlflow.log_param("Date A", date_a)
         mlflow.log_param("Date B", date_b)
         calculate_score(date_a, date_b)
+        calculate_metrics(date_a,date_b)
         mlflow.end_run()
-
+
```
Comment (durandom) [inline metric_tracking/__init__.py]: @tumido do we have a bot commenting on style? If so, can we get it enabled in this repo?

File: Makefile [hunk 2]
```diff
@@ -8,12 +8,14 @@ oc_build_head:
 	oc start-build systems-clustering --from-archive build/HEAD.tar.gz --follow
 
 oc_cluster_train:
-	oc new-app --template systems-clustering-job \
-		--param AIOPS_TRAINING_DATE=${AIOPS_TRAINING_DATE} \
-		--param CEPH_KEY=${CEPH_KEY} \
-		--param CEPH_SECRET=${CEPH_SECRET} \
-		--param CEPH_ENDPOINT=${CEPH_ENDPOINT} \
-		--param CEPH_BUCKET=${CEPH_BUCKET}
+	oc new-app mlflow-experiment-job --param APP_IMAGE_URI=your-application-image-name\
```
Comment (durandom) [inline Makefile]: I think training and running the experiment are two different things.

The training would create a new model and store it in ceph.

The experiment works either on previous models, or creates a new one and reports some metrics. 
I think the current metric code only compares two clusters, which is ok for now.

So, I'd create a new task: `oc_run_experiment` and keep the `oc_cluster_train` task as is

File: README.md [hunk 3]
```diff
@@ -1,31 +1,19 @@
 # Clustering Systems
 
 ## Running the clustering on OpenShift
```
Comment (durandom) [inline README.md]: again, keep this section as is and create a new section for the experiment.

File: Makefile (hunk 2 above)
Comment (hemajv) [reply Makefile]: Sure, that makes sense. I have now created a separate task for this.

File: metric_tracking/__init__.py (hunk 1 above)
Comment (tumido) [reply metric_tracking/__init__.py]: @durandom We currently use a Travis for that. One step in its job is to do the linting. That's what you mean? Sure, we can enable it. :) It you'd like to integrate a bot pointing out the violations directly in a comment, we currently don't have that enabled/available, but I think AI-CoE bots do something like that, don't they?

Comment (durandom) [reply metric_tracking/__init__.py]: can't miq-bot do python? @juliancheal ?
@goern - do your linters post comments in the PR?

File: README.md (hunk 3 above)
Comment (hemajv) [reply README.md]: Done :)

File: README.md [hunk 4]
```diff
@@ -78,6 +78,82 @@ Name: cluster, Length: 40162, dtype: int32
...
+Once you have the code on openshift, you can have multiple instances of your experiment running with different parameters.
+
+## First Stage: Obtaining the application code on OpenShift
+
+To run this application on openshift, we need to create a container image for it. In other words, this means that we need to download the code that we want to run on openshift. To do this a image build template should already be available in your openshift namespace.
+
+To download the source code for our experiment to openshift, use the following command:
+
+```
+    oc process mlflow-experiment-bc --param APPLICATION_NAME=your-application-name --param GIT_URI=https://github.com/ManageIQ/aiops-insights-clustering.git --param APP_FILE=app.py | oc create -f -
```
[the remark this answers is not in the document]
Comment (hemajv) [reply README.md]: Yes that's a good point, the template by default selects the app.py file so I have removed this parameter.

File: README.md [hunk 5]
```diff
@@ -78,6 +78,82 @@ Name: cluster, Length: 40162, dtype: int32
...
+```
+
+## Second Stage: Running an experiment
+
+After the image is built, we can use it to run an experiment on OpenShift.
+
+Edit the following line in the Makefile to set the APP_IMAGE_URI to your application image as follows:
+
+```
+    oc new-app mlflow-experiment-job --param APP_IMAGE_URI=your-application-image-name
```
[the remark this answers is not in the document]
Comment (hemajv) [reply README.md]: corrected!

**Pull request merged**